llvm/lib/Transforms/Scalar/SROA.cpp

```diff
@@ -3937,30 +3937,63 @@ class AggLoadStoreRewriter : public InstVisitor<AggLoadStoreRewriter, bool> {
     return false;
   }
 
-  // Fold gep (select cond, ptr1, ptr2) => select cond, gep(ptr1), gep(ptr2)
+  // Fold gep (select cond, ptr1, ptr2), idx
+  //   => select cond, gep(ptr1, idx), gep(ptr2, idx)
+  // and  gep ptr, (select cond, idx1, idx2)
+  //   => select cond, gep(ptr, idx1), gep(ptr, idx2)
   bool foldGEPSelect(GetElementPtrInst &GEPI) {
-    if (!GEPI.hasAllConstantIndices())
-      return false;
+    // Check whether the GEP has exactly one select operand and all indices
+    // will become constant after the transform.
+    auto IsValidOp = [](Value *Op) {
+      return Op->getType()->isPointerTy() || isa<ConstantInt>(Op);
+    };
+
+    SelectInst *Sel = nullptr;
+    for (Value *Op : GEPI.operands()) {
+      if (auto *SI = dyn_cast<SelectInst>(Op)) {
```

Comment:
Do we want to restrict it to selects of known integer values? Allowing all selects would result in splitting into GEPs that SROA would still not be able to see through.
Select of constants is always useful.
Select of one const, is partially useful.
Select of non-const is a gamble some may eventually evaluate to a known value, some would not.

Reply:
In the context of SROA, I believe only the "all constant" case is useful. If you have one access via a non-constant GEP, that will always prevent splitting and promotion.

Reply:
> I believe only the "all constant" case is useful. 

Counter-example: If we had `gep(select(1,select(2,3)))`, it could be further split into `gep(const)` when we'd visited the split geps after the first iteration. 

`isa<ConstantInt>` is too conservative. I think the correct check here is whether LLVM can compute a known value. for each branch of select, as the select branches may be other ops, yet could be evaluated to an int value.

Reply:
Interesting case. I agree that we could do the transform nested selects if their leaves are constant. I've added a test for this, but left the logic alone -- this seems like something we should evaluate separately.

Reply:
Ack. We can handle it later.